### ISSUE TITLE
Why Adaptive learning rate scheduler is not supported with dist kvstore

### ISSUE THREAD

**Issue by threeleafzerg** [awslabs/sockeye#482]
I saw this from the sockeye source code:

>             utils.check_condition(
                not issubclass(type(self.optimizer_config.lr_scheduler),
                               lr_scheduler.AdaptiveLearningRateScheduler),
                "Adaptive learning rate schedulers not supported with a dist kvstore. "
                "Try a fixed schedule such as %s." % C.LR_SCHEDULER_FIXED_RATE_INV_SQRT_T)

Is it the limitation from mxnet framework or sockeye itself? 
Thanks!

**Comment by fhieber** [comment]
The AdaptiveLearningRateSchedulers available in Sockeye use a method called `new_evaluation_result(has_improved: bool)` which controls the learning rate based on whether the optimized metric did improve on the validation set at the current checkpoint. 
With distributed training, you have multiple training processes running independently on multiple hosts. If you use a standard Sockeye training processes as workers, each of these processes run their own validation iterator on each checkpoint and may set `has_improved` differently.
Moreover, to my knowledge, there is no way to set the learning rate of the global optimizer that is pickled and sent to the MXNet distributed training server. 
But I am not too familiar with this part of MXNet, so maybe this has changed by now.

Anyway, for multi-host Sockeye training runs, one would need to define a master process that runs validation at each checkpoint and controls the global learning rate based on that. All other workers should skip this.

**Comment by threeleafzerg** [comment]
Thanks fhieber for your quick response. 
I already enabled multi host training in my local repo. Currently I am checking the accuracy of multi host training. 
From your comment, it looks like it's simply a limitation from sockeye. 
For "has_improved":
1) In my solution, I didn't divide validation set into all nodes. (all nodes run the same validation set.)  When compared to running validation and checkpoint only on master node, the total time is the same. But the logic is quite simple. 
2) In my solution, I only support sync dist kvstore, so has_improved should be same in all nodes. 

But I don't get at "there is no way to set the learning rate of the global optimizer that is pickled and sent to the MXNet distributed training server". Do you mean learning rate of the global optimizer need to be synced up across all nodes?

**Comment by tdomhan** [comment]
AdaptiveLearningRateSchedulers is a custom scheduler implemented in Sockeye. It relies on having a single object which gets informed about validation metric scores. With distributed training the scheduler is pickled and sent remotely. On the remote host no one would actually inform the scheduler about new evaluation metrics. So for distributed training it's best to simply pick a different learning rate scheduler (e.g. a fixed decrease in the learning rate). I hope that makes sense and helps.

**Comment by tdomhan** [comment]
closing for now. Let us know if you have additional questions.
